File: Core/YABACore/src/commonMain/kotlin/dev/subfly/yabacore/state/creation/docmark/DocmarkCreationStateMachine.kt
package dev.subfly.yabacore.state.creation.docmark

import dev.subfly.yabacore.common.IdGenerator
import dev.subfly.yabacore.filesystem.DocmarkFileManager
import dev.subfly.yabacore.filesystem.access.YabaFileAccessor
import dev.subfly.yabacore.managers.AllBookmarksManager
import dev.subfly.yabacore.managers.DocmarkManager
import dev.subfly.yabacore.managers.FolderManager
import dev.subfly.yabacore.managers.TagManager
import dev.subfly.yabacore.model.utils.BookmarkAppearance
import dev.subfly.yabacore.model.utils.BookmarkKind
import dev.subfly.yabacore.model.utils.CardImageSizing
import dev.subfly.yabacore.model.utils.DocmarkType
import dev.subfly.yabacore.preferences.SettingsStores
import dev.subfly.yabacore.state.base.BaseStateMachine
import dev.subfly.yabacore.webview.WebShellLoadResult
import io.github.vinceglb.filekit.name
import io.github.vinceglb.filekit.readBytes
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.IO
import kotlinx.coroutines.withContext

class DocmarkCreationStateMachine :
    BaseStateMachine<DocmarkCreationUIState, DocmarkCreationEvent>(
        initialState = DocmarkCreationUIState(),
    ) {
    private var isInitialized = false
    private val preferencesStore
        get() = SettingsStores.userPreferences

    override fun onEvent(event: DocmarkCreationEvent) {
        when (event) {
            is DocmarkCreationEvent.OnInit -> onInit(event)
            DocmarkCreationEvent.OnPickDocument -> onPickDocument()
            is DocmarkCreationEvent.OnDocumentFromShare -> onDocumentFromShare(event)
            DocmarkCreationEvent.OnClearDocument -> onClearDocument()
            DocmarkCreationEvent.OnCyclePreviewAppearance -> onCyclePreviewAppearance()
            is DocmarkCreationEvent.OnDocumentMetadataExtracted -> onDocumentMetadataExtracted(event)
            is DocmarkCreationEvent.OnSetGeneratedPreview -> onSetGeneratedPreview(event)
            is DocmarkCreationEvent.OnChangeLabel -> onChangeLabel(event)
            is DocmarkCreationEvent.OnChangeDescription -> onChangeDescription(event)
            is DocmarkCreationEvent.OnChangeSummary -> onChangeSummary(event)
            is DocmarkCreationEvent.OnSelectFolder -> onSelectFolder(event)
            is DocmarkCreationEvent.OnSelectTags -> onSelectTags(event)
            is DocmarkCreationEvent.OnSave -> onSave(event)
            is DocmarkCreationEvent.OnWebInitialContentLoad -> onWebInitialContentLoad(event)
        }
    }

    private fun onInit(event: DocmarkCreationEvent.OnInit) {
        if (isInitialized) return
        isInitialized = true

        launch {
            val preferences = preferencesStore.get()
            updateState {
                it.copy(
                    bookmarkAppearance = preferences.preferredBookmarkAppearance,
                    cardImageSizing = preferences.preferredCardImageSizing,
                )
            }

            event.docmarkIdString?.let { docmarkId ->
                val existing = DocmarkManager.getDocmarkDetail(docmarkId)
                if (existing != null) {
                    updateState {
                        it.copy(
                            label = existing.label,
                            description = existing.description ?: "",
                            summary = existing.summary ?: "",
                            metadataTitle = existing.metadataTitle,
                            metadataDescription = existing.metadataDescription,
                            metadataAuthor = existing.metadataAuthor,
                            metadataDate = existing.metadataDate,
                            metadataIdentifier = existing.metadataIdentifier,
                            selectedFolder = existing.parentFolder,
                            selectedTags = existing.tags,
                            editingDocmark = existing,
                            docmarkType = existing.docmarkType,
                        )
                    }
                }
            }

            event.initialFolderId?.let { folderId ->
                val folder = FolderManager.getFolder(folderId)
                if (folder != null) updateState { it.copy(selectedFolder = folder) }
            }

            event.initialTagIds?.let { tagIds ->
                val tags = tagIds.mapNotNull { tagId -> TagManager.getTag(tagId) }
                if (tags.isNotEmpty()) updateState { it.copy(selectedTags = tags) }
            }

            if (currentState().selectedFolder == null) {
                val temporaryUncategorized = FolderManager.createUncategorizedFolderModel()
                updateState {
                    it.copy(
                        selectedFolder = temporaryUncategorized,
                        uncategorizedFolderCreationRequired = true,
                    )
                }
            }
        }
    }

    private fun onPickDocument() {
        if (currentState().isInEditMode) return
        launch {
            updateState { it.copy(isLoading = true, error = null) }
            try {
                val file =
                    YabaFileAccessor.pickSingleFile(extensions = listOf("pdf", "epub")) ?: run {
                        updateState { it.copy(isLoading = false) }
                        return@launch
                    }
                val bytes =
                    withContext(Dispatchers.IO) {
                        file.readBytes()
                    }
                val sourceName = file.name
                val docmarkType = inferDocmarkTypeFromFileName(sourceName)
                if (docmarkType == null) {
                    updateState {
                        it.copy(
                            isLoading = false,
                            error = DocmarkCreationError.DocumentReadFailed,
                        )
                    }
                    return@launch
                }
                updateState {
                    it.copy(
                        documentBytes = bytes,
                        docmarkType = docmarkType,
                        sourceFileName = sourceName,
                        metadataTitle = null,
                        metadataDescription = null,
                        metadataAuthor = null,
                        metadataDate = null,
                        metadataIdentifier = null,
                        isLoading = true,
                        error = null,
                    )
                }
            } catch (_: Exception) {
                updateState {
                    it.copy(
                        isLoading = false,
                        error = DocmarkCreationError.DocumentReadFailed,
                    )
                }
            }
        }
    }

    private fun inferDocmarkTypeFromFileName(fileName: String): DocmarkType? {
        val ext = fileName.substringAfterLast('.', "").lowercase()
        return when (ext) {
            "pdf" -> DocmarkType.PDF
            "epub" -> DocmarkType.EPUB
            else -> null
        }
    }

    private fun onDocumentFromShare(event: DocmarkCreationEvent.OnDocumentFromShare) {
        if (currentState().isInEditMode) return
        val sourceFileName = event.sourceFileName?.trim()?.ifBlank { null }
        updateState { state ->
            state.copy(
                documentBytes = event.bytes,
                docmarkType = event.docmarkType,
                sourceFileName = sourceFileName,
                metadataTitle = null,
                metadataDescription = null,
                metadataAuthor = null,
                metadataDate = null,
                metadataIdentifier = null,
                previewImageBytes = null,
                previewImageExtension = "png",
                isLoading = true,
                error = null,
            )
        }
    }

    private fun onClearDocument() {
        if (currentState().isInEditMode) return
        updateState {
            it.copy(
                documentBytes = null,
                docmarkType = null,
                sourceFileName = null,
                metadataTitle = null,
                metadataDescription = null,
                metadataAuthor = null,
                metadataDate = null,
                metadataIdentifier = null,
                previewImageBytes = null,
                isLoading = false,
                error = null,
            )
        }
    }

    private fun onCyclePreviewAppearance() {
        val state = currentState()
        val (nextAppearance, nextSizing) = when (state.bookmarkAppearance) {
            BookmarkAppearance.LIST -> BookmarkAppearance.CARD to CardImageSizing.SMALL
            BookmarkAppearance.CARD -> when (state.cardImageSizing) {
                CardImageSizing.SMALL -> BookmarkAppearance.CARD to CardImageSizing.BIG
                CardImageSizing.BIG -> BookmarkAppearance.GRID to state.cardImageSizing
            }
            BookmarkAppearance.GRID -> BookmarkAppearance.LIST to state.cardImageSizing
        }
        updateState {
            it.copy(
                bookmarkAppearance = nextAppearance,
                cardImageSizing = nextSizing,
            )
        }
    }

    private fun onDocumentMetadataExtracted(event: DocmarkCreationEvent.OnDocumentMetadataExtracted) {
        updateState {
            it.copy(
                metadataTitle = event.metadataTitle,
                metadataDescription = event.metadataDescription,
                metadataAuthor = event.metadataAuthor,
                metadataDate = event.metadataDate,
                metadataIdentifier = event.metadataIdentifier,
            )
        }
    }

    private fun onSetGeneratedPreview(event: DocmarkCreationEvent.OnSetGeneratedPreview) {
        updateState {
            it.copy(
                previewImageBytes = event.imageBytes,
                previewImageExtension = event.extension.lowercase().removePrefix(".").ifBlank { "png" },
            )
        }
    }

    private fun onWebInitialContentLoad(event: DocmarkCreationEvent.OnWebInitialContentLoad) {
        updateState {
            it.copy(
                isLoading = false,
                error = if (event.result == WebShellLoadResult.Error) DocmarkCreationError.PreviewExtractionFailed else null,
            )
        }
    }

    private fun onChangeLabel(event: DocmarkCreationEvent.OnChangeLabel) {
        updateState { it.copy(label = event.newLabel) }
    }

    private fun onChangeDescription(event: DocmarkCreationEvent.OnChangeDescription) {
        updateState { it.copy(description = event.newDescription) }
    }

    private fun onChangeSummary(event: DocmarkCreationEvent.OnChangeSummary) {
        updateState { it.copy(summary = event.newSummary) }
    }

    private fun onSelectFolder(event: DocmarkCreationEvent.OnSelectFolder) {
        updateState {
            it.copy(
                selectedFolder = event.folder,
                uncategorizedFolderCreationRequired = false,
            )
        }
    }

    private fun onSelectTags(event: DocmarkCreationEvent.OnSelectTags) {
        updateState { it.copy(selectedTags = event.tags) }
    }

    private fun onSave(event: DocmarkCreationEvent.OnSave) {
        val state = currentState()
        var selectedFolder = state.selectedFolder ?: return

        if (!state.isInEditMode && state.documentBytes == null) {
            updateState { it.copy(error = DocmarkCreationError.NoDocument) }
            event.onErrorCallback()
            return
        }

        if (!state.isInEditMode && state.docmarkType == null) {
            updateState { it.copy(error = DocmarkCreationError.NoDocument) }
            event.onErrorCallback()
            return
        }

        launch {
            updateState { it.copy(isSaving = true, error = null) }

            try {
                if (state.uncategorizedFolderCreationRequired) {
                    selectedFolder = FolderManager.ensureUncategorizedFolderVisible()
                }

                val bookmarkId = state.editingDocmark?.id ?: IdGenerator.newId()
                val label = state.label.ifBlank { "Document" }
                if (state.editingDocmark != null) {
                    AllBookmarksManager.updateBookmarkMetadata(
                        bookmarkId = bookmarkId,
                        folderId = selectedFolder.id,
                        kind = BookmarkKind.FILE,
                        label = label,
                        description = state.description.ifBlank { null },
                        tagIds = state.selectedTags.map { it.id },
                    )

                    val existingTagIds = state.editingDocmark.tags.map { it.id }.toSet()
                    val newTagIds = state.selectedTags.map { it.id }.toSet()
                    state.selectedTags.filter { it.id !in existingTagIds }.forEach { tag ->
                        AllBookmarksManager.addTagToBookmark(tag.id, bookmarkId)
                    }
                    state.editingDocmark.tags.filter { it.id !in newTagIds }.forEach { tag ->
                        AllBookmarksManager.removeTagFromBookmark(tag.id, bookmarkId)
                    }
                } else {
                    AllBookmarksManager.createBookmarkMetadata(
                        id = bookmarkId,
                        folderId = selectedFolder.id,
                        kind = BookmarkKind.FILE,
                        label = label,
                        description = state.description.ifBlank { null },
                        tagIds = state.selectedTags.map { it.id },
                        previewImageBytes = state.previewImageBytes,
                        previewImageExtension = state.previewImageExtension,
                        previewIconBytes = null,
                    )
                    val docBytes = state.documentBytes ?: error("Missing document bytes")
                    val docType = state.docmarkType ?: error("Missing docmark type")
                    DocmarkFileManager.saveDocumentBytes(
                        bookmarkId = bookmarkId,
                        bytes = docBytes,
                        type = docType,
                    )
                }

                DocmarkManager.createOrUpdateDocDetails(
                    bookmarkId = bookmarkId,
                    summary = state.summary.ifBlank { null },
                    docmarkType = if (state.editingDocmark != null) null else state.docmarkType,
                    metadataTitle = state.metadataTitle,
                    metadataDescription = state.metadataDescription,
                    metadataAuthor = state.metadataAuthor,
                    metadataDate = state.metadataDate,
                    metadataIdentifier = state.metadataIdentifier,
                )

                updateState { it.copy(isSaving = false, error = null) }
                event.onSavedCallback()
            } catch (_: Exception) {
                updateState {
                    it.copy(
                        isSaving = false,
                        error = DocmarkCreationError.SaveFailed,
                    )
                }
                event.onErrorCallback()
            }
        }
    }

    override fun clear() {
        isInitialized = false
        super.clear()
    }
}

File: Darwin/YABACore/State/Creation/Docmark/DocmarkCreationStateMachine.swift
//
//  DocmarkCreationStateMachine.swift
//  YABACore
//

import Foundation
import SwiftUI

@MainActor
public final class DocmarkCreationStateMachine: YabaBaseObservableState<DocmarkCreationUIState>, YabaScreenStateMachine {
    public override init(initialState: DocmarkCreationUIState = DocmarkCreationUIState()) {
        super.init(initialState: initialState)
    }

    public func send(_ event: DocmarkCreationEvent) async {
        switch event {
        case let .onInit(id, folderId, tagIds, uncategorizedFolderCreationRequired):
            apply {
                $0.editingBookmarkId = id
                $0.selectedFolderId = folderId
                $0.selectedTagIds = tagIds ?? []
                if id == nil {
                    $0.uncategorizedFolderCreationRequired = uncategorizedFolderCreationRequired
                } else {
                    $0.uncategorizedFolderCreationRequired = false
                }
            }
        case .onPickDocument:
            break
        case .onClearDocument:
            apply {
                $0.pickedDocumentData = nil
                $0.sourceFileName = nil
                $0.docmarkType = nil
                $0.metadataTitle = nil
                $0.metadataDescription = nil
                $0.metadataAuthor = nil
                $0.metadataDate = nil
                $0.previewImageData = nil
                $0.isLoading = false
                $0.lastError = nil
            }
        case let .onDocumentFromShare(data, name, type):
            apply {
                $0.pickedDocumentData = data
                $0.sourceFileName = name
                $0.docmarkType = type
                $0.metadataTitle = nil
                $0.metadataDescription = nil
                $0.metadataAuthor = nil
                $0.metadataDate = nil
                $0.previewImageData = nil
                $0.isLoading = true
                $0.lastError = nil
            }
        case .onCyclePreviewAppearance:
            apply {
                switch $0.bookmarkAppearance {
                case .list: $0.bookmarkAppearance = .card
                case .card: $0.bookmarkAppearance = .grid
                case .grid: $0.bookmarkAppearance = .list
                }
            }
        case let .onDocumentMetadataExtracted(title, desc, author, date):
            apply {
                $0.metadataTitle = title
                $0.metadataDescription = desc
                $0.metadataAuthor = author
                $0.metadataDate = date
            }
        case let .onSetGeneratedPreview(data, _):
            apply { $0.previewImageData = data }
        case let .onChangeLabel(s):
            apply { $0.label = s }
        case let .onChangeDescription(s):
            apply { $0.bookmarkDescription = s }
        case .onApplyFromMetadata:
            withAnimation {
                apply {
                    if let t = $0.metadataTitle, !t.isEmpty { $0.label = t }
                    if let d = $0.metadataDescription, !d.isEmpty { $0.bookmarkDescription = d }
                }
            }
        case let .onSelectFolderId(id):
            apply {
                $0.selectedFolderId = id
                $0.uncategorizedFolderCreationRequired = false
            }
        case let .onSelectTagIds(ids):
            apply { $0.selectedTagIds = ids }
        case let .onWebInitialContentLoad(result):
            apply {
                $0.isLoading = false
                if result == .error {
                    $0.lastError = "Preview extraction failed"
                } else {
                    $0.lastError = nil
                }
            }
        case .onDocumentExtractionFinished:
            apply { $0.isLoading = false }
        case .onSave:
            await persist()
        case .onTogglePinned:
            apply { $0.isPinned.toggle() }
        case let .create(bookmarkId, folderId, label, bookmarkDescription, isPinned, tagIds):
            AllBookmarksManager.queueCreateBookmark(
                bookmarkId: bookmarkId,
                folderId: folderId,
                kind: .file,
                label: label,
                bookmarkDescription: bookmarkDescription,
                isPinned: isPinned,
                tagIds: tagIds
            )
            DocmarkManager.queueEnsureInitialDocDetail(bookmarkId: bookmarkId)
        }
    }

    private func persist() async {
        let folderId = state.selectedFolderId
        let trimmedLabel = state.label.trimmingCharacters(in: .whitespacesAndNewlines)
        let label = trimmedLabel.isEmpty ? "Document" : trimmedLabel
        guard let folderId, !folderId.isEmpty else {
            apply { $0.lastError = "Folder required" }
            return
        }
        if state.editingBookmarkId == nil {
            guard state.pickedDocumentData != nil, state.docmarkType != nil else {
                apply { $0.lastError = "No document selected" }
                return
            }
        }
        if state.uncategorizedFolderCreationRequired {
            do {
                try await CoreOperationQueue.shared.queueAndAwait(name: "EnsureUncategorizedFolderVisible") { context in
                    try FolderManager.ensureUncategorizedFolderVisibleInContext(context)
                }
                apply { $0.uncategorizedFolderCreationRequired = false }
            } catch {
                apply {
                    $0.lastError = String(describing: error)
                    $0.isSaving = false
                }
                return
            }
        }
        apply { $0.lastError = nil; $0.isSaving = true }
        let bid = state.editingBookmarkId ?? UUID().uuidString
        let summary = state.summary.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
        if state.editingBookmarkId != nil {
            AllBookmarksManager.queueUpdateBookmarkMetadata(
                bookmarkId: bid,
                folderId: folderId,
                kind: .file,
                label: label,
                bookmarkDescription: state.bookmarkDescription.nilIfEmpty,
                isPinned: state.isPinned,
                tagIds: state.selectedTagIds
            )
        } else {
            AllBookmarksManager.queueCreateBookmark(
                bookmarkId: bid,
                folderId: folderId,
                kind: .file,
                label: label,
                bookmarkDescription: state.bookmarkDescription.nilIfEmpty,
                isPinned: state.isPinned,
                tagIds: state.selectedTagIds
            )
            AllBookmarksManager.queueSetBookmarkPreviewAssets(
                bookmarkId: bid,
                imageBytes: state.previewImageData,
                iconBytes: nil
            )
        }
        DocmarkManager.queueCreateOrUpdateDocDetails(
            bookmarkId: bid,
            summary: summary,
            docmarkType: state.editingBookmarkId != nil ? nil : state.docmarkType,
            metadataTitle: state.metadataTitle,
            metadataDescription: state.metadataDescription,
            metadataAuthor: state.metadataAuthor,
            metadataDate: state.metadataDate
        )
        if state.editingBookmarkId == nil, let docBytes = state.pickedDocumentData {
            DocmarkManager.queueUpsertDocBookmarkPayloadBytes(bookmarkId: bid, documentBytes: docBytes)
        }
        apply { $0.isSaving = false }
    }
}
